web/src/main/webapp/my-app/layout/controllers.js

```diff
@@ -267,6 +357,120 @@ define(['angular', 'jquery'], function(angular, $) {
       // Inherit from BaseWidgetFunctionsController
       $controller('BaseWidgetFunctionsController',
       {$scope: $scope, childController: vm});
+      /**
+       * Log whenever a widget is moved
+       * @param eventType {String} Kind of interaction that moved the widget
+       * @param fname {String} The moved widget's fname
+       * @param dropIndex {Number} Index where the widget landed
+       * @param startIndex {Number} (optional) Index before moving the widget
+       * @return {boolean}
+       */
+      $scope.logEvent = function(eventType, fname, dropIndex, startIndex) {
+        switch (eventType) {
+          case 'dragEnd':
+            $log.info('Dragged ' + fname + ' to index ' + dropIndex);
+            break;
+          case 'keyboardMove':
+            $log.info('Moved ' + fname + 'from index ' + startIndex +
+              ' to index ' + dropIndex);
+            break;
+          default:
+            return true;
+        }
+      };
+
+      /**
+       * Respond to arrow key-presses when focusing a movable list element
+       * @param widget {Object} The widget trying to move
+       * @param event {Object} The event object
+       * @return {boolean} True if proper conditions aren't met
+       */
+      $scope.moveWithKeyboard = function(widget, event) {
```

Comment:
How is this different from `web/src/main/webapp/my-app/layout/controllers.js` line 79-130? are both needed?

Reply:
Not necessarily, no. It's duplicated because the two views use different controllers, and refactoring those controllers isn't in the scope of this story. It's something we should do at some point.